### PR TITLE
[#674] improvement(server): use ByteString#asReadOnlyByteBuffer to reduce the memory allocation

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

Use ByteString#asReadOnlyByteBuffer to reduce the memory allocation in GRPC of `sendShuffleData`

### Why are the changes needed?

Fix: #674 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

### CODE REVIEW

**Comment by zuston** [comment]
PTAL @jerqi @advancedxy

**Comment by zuston** [comment]
> Have you done some benchmark tests?

No. But I tried this patch to run some jobs, and from the flamegraph, the hotspot of memory allocation don't occur.

File: storage/src/main/java/org/apache/uniffle/storage/handler/impl/LocalFileWriter.java [hunk 1]
```diff
@@ -31,15 +33,26 @@ public class LocalFileWriter implements FileWriter, Closeable {
...
   private DataOutputStream dataOutputStream;
   private FileOutputStream fileOutputStream;
+  private FileChannel fileChannel;
   private long nextOffset;
 
   public LocalFileWriter(File file) throws IOException {
     fileOutputStream = new FileOutputStream(file, true);
+    fileChannel = fileOutputStream.getChannel();
     // init fsDataOutputStream
     dataOutputStream = new DataOutputStream(new BufferedOutputStream(fileOutputStream));
```
[the remark this answers is not in the document]
Comment (zuston) [reply storage/src/main/java/org/apache/uniffle/storage/handler/impl/LocalFileWriter.java:43]: Emm... Actually, the `FileWriter` interface is not only for writing index data but also writing shuffle data, which is split with 2 methods, but they are exclusive. So only one of `dataOutputStream` and `fileChannel` will be used.

**Comment by jerqi** [comment]
> > Have you done some benchmark tests?
> 
> No. But I tried this patch to run some jobs, and from the flamegraph, the hotspot of memory allocation don't occur.

We would better have some benchmark tests.

**Comment by zuston** [comment]
> > > Have you done some benchmark tests?
> > 
> > 
> > No. But I tried this patch to run some jobs, and from the flamegraph, the hotspot of memory allocation don't occur.
> 
> We would better have some benchmark tests.

It's OK. Let me do some tests later.

**Comment by smallzhongfeng** [comment]
https://github.com/apache/incubator-uniffle/pull/742 has been merged, let's go ahead.

**Comment by zuston** [comment]
It looks `ShufflePartitionedBlock` should also be updated in netty implementation.

**Comment by jerqi** [comment]
You can see https://github.com/apache/incubator-uniffle/issues/586

**Comment by zuston** [comment]
> > It looks `ShufflePartitionedBlock` should also be updated in netty implementation.
> 
> So, this pr should be delay after some other netty prs checked in?

I think yes. And I'm evaluating whether we need this pr.

**Pull request closed without merging**